sklearn/preprocessing/_discretization.py

```diff
@@ -231,32 +244,70 @@ def fit(self, X, y=None, sample_weight=None):
 
         n_samples, n_features = X.shape
 
-        if sample_weight is not None and self.strategy == "uniform":
-            raise ValueError(
-                "`sample_weight` was provided but it cannot be "
-                "used with strategy='uniform'. Got strategy="
-                f"{self.strategy!r} instead."
-            )
-
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it is important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
             )
-
-        n_features = X.shape[1]
-        n_bins = self._validate_n_bins(n_features)
+            # Since we already used the weights when resampling when provided,
+            # we set them back to `None` to avoid accounting for the weights twice
+            # in subsequent operations to compute weight-aware bin edges with
+            # quantiles or k-means.
+            sample_weight = None
 
         if sample_weight is not None:
             sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
 
+        n_features = X.shape[1]
+        n_bins = self._validate_n_bins(n_features)
+
         bin_edges = np.zeros(n_features, dtype=object)
+
+        # TODO(1.9): remove and switch to quantile_method="averaged_inverted_cdf"
+        # by default.
+        if self.strategy == "quantile" and quantile_method == "warn":
```

Comment:
So here we should use a local variable, initialized from the constructor parameter.

```suggestion
        quantile_method = self.quantile_method
        if self.strategy == "quantile" and quantile_method == "warn":
```